llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -365,6 +365,80 @@ void groupEntryPoints(const Module &M, EntryPointGroupMap &EntryPointsGroups,
     EntryPointsGroups[GLOBAL_SCOPE_NAME] = {};
 }
 
+// For device global variables with the 'device_image_scope' property,
+// the function checks that there are no usages of a single device global
+// variable from kernels grouped to different modules.
+void checkImageScopedDeviceGlobals(const Module &M,
+                                   const EntryPointGroupMap &GMap) {
+  // Early exit if there is only one group
+  if (GMap.size() < 2)
+    return;
+
+  // Reverse the EntryPointGroupMap to get a map of entry point -> module's name
+  unsigned EntryPointNumber = 0;
+  for (const auto &Group : GMap) {
+    EntryPointNumber += static_cast<unsigned>(Group.second.size());
+  }
+  DenseMap<const Function *, const StringRef *> EntryPointModules(
+      EntryPointNumber);
+  for (const auto &Group : GMap) {
+    auto *ModuleName = &Group.first;
+    for (const auto *F : Group.second) {
+      EntryPointModules.insert({F, ModuleName});
+    }
+  }
```

Comment:
The suggested change implements a little bit different logic. I'm accumulating `EntryPointNumber` as a total number of entry points in the `GMap` and then using the value to reserve the right amount of memory for the `EntryPointModules` map. This is the reason why `GMap` in my example is traversed twice. Exactly, it can look as a premature optimization, but I believe there is a trade off between an extra traverse and many allocations in the `DenseMap`.